[agw][liagentd] liagentd service implementation

ymasmoudi opened the pull request:
## Summary

LIAgentD implements LI X3 interface as specified in ETSI 103 221-2. 
An nprobe task representing a warrant must be configured through swagger and will contain informations like task ID (uuid v4), Target ID (Imsi, Imei, ...),  Target Type, Correlation ID (used to correlate X2 and X3 records),...
   
Each configured task in swagger will be translated to,
- pipelined mconfig : LiUes (list of subscriber IDs to be intercepted)
- liagentd mconfig : list of nprobe tasks details

\
First, PipelineD will mirror all the traffic of the targeted subscriber to a dedicated network interface (li_port)
Then, LiAgentD will listen continuously on li_port and process each packet as follow,
- For each new target, interrogates MobilityD to retrieve the subscriber ID from IP address
- Create a new intercept state (stored in hash map for now) 
- Create X3 record by encapsulating the mirrored packet (starting from IP layer) in X3 header as defined in ETSI 103 221-2.
- Exports the record over TLS (certificates configured in /etc/magma/liagentd.yml)

\
X3 Header
```
typedef struct {
  uint16_t version;
  uint16_t pdu_type;
  uint32_t header_length;
  uint32_t payload_length;
  uint16_t payload_format;
  uint16_t payload_direction;
  uint8_t xid[XID_LENGTH];
  uint64_t correlation_id;
  ConditionalAttributes attrs;   // TLV fields - timestamp,  sequence number, target id, NFID
} __attribute__((__packed__)) X3Header;
```
\
In addition, LiAgentD uses two timers,
- sync_interval : Interval between mconfig loading in memory to retrieve latest tasks (by default 60 seconds)
- inactivity_time: time after which an intercept state is deleted if no packet is exported
\
An intercept state contains the following information,

```
typedef struct {
  std::string task_id;
  std::string target_id;
  std::string domain_id;
  uint64_t last_exported;
  uint64_t correlation_id;
  uint64_t sequence_number;
} InterceptState;
```

In addition, this PR implements as well,
- support retries and establishment of tls session when it is down.
- support state cleanup when traffic is being processed

## Test Plan

3 Unit tests were implemented to cover non IP packet/malformed, unknown subscriber and a valid use case,
```
ninja -C   /home/vagrant/build/c/li_agent
ninja: Entering directory `/home/vagrant/build/c/li_agent'
[21/21] Linking CXX executable liagentd
cd  /home/vagrant/build/c/li_agent && ctest --output-on-failure
Test project /home/vagrant/build/c/li_agent
    Start 1: test_pdu_generator
1/1 Test #1: test_pdu_generator ...............   Passed    0.03 sec

100% tests passed, 0 tests failed out of 1
```


Fully validated using integration tests and remote server (https://github.com/ymasmoudi/nprobe-collector)
Traffic can be generated using the integration testcase, 
- make integ_test TESTS=s1aptests/test_attach_detach_attach_ul_tcp_data.py 

```
X3 PDU Record (IPv4 Packet) for XID 29f28e1c-f230-486a-a860-f5a784ab9174, ID 2225836404, CIN 1116605273, CorrID 0x0866cb39 790844a8 received from ETSI 103 221
Type:                                    X3 PDU (35)
SR Type:                                 IPv4 Packet
Version:         2
PDU Type:        X3 PDU (2)
Header Length:   64
Payload Length:  52
Payload Format:  IPv4 Packet (5)
Direction:       from-target (3)
XID:             29f28e1c-f230-486a-a860-f5a784ab9174
CaseId/LITID:    2225836404
CorrelationId:   0866cb39 790844a8
CIN:             1116605273
Sequence Number: 10
TimeStamp UNIX Time:                     Seconds since 01.01.1970: 1622027236 (0x60ae2be4), micro seconds: 802505 (0x000c3ec9)
  Local time:                            2021-05-26 13:07:16, [+02:00], [summer time(2)], fraction 803 ms
Attribute Timestamp (9):
Attribute length: 8
Timestamp:       seconds since epoc 0, nano seconds 1621857809
Timestamp UNIX Time:                     Seconds since 01.01.1970: 1622027236 (0x60ae2be4), micro seconds: 802515 (0x000c3ed3)
  Local time:                            2021-05-26 13:07:16, [+02:00], [summer time(2)], fraction 803 ms
Attribute Sequence Number (8):
Attribute length: 8
Sequence Number: 10
IP Header
    Version:                             IPv4 (4)
    Header Length:                       20 Bytes
    Type of Service:                     0x00
        Precedence:                      Routine (0)
        D(elay)                          0 = Normal Delay
        T(roughput)                      0 = Normal Throughput
        R(elibility)                     0 = Normal Relibility
    Total Length:                        52
    Identification:                      0x2fe4
    Flags:                               0x02
        DF:                              0 = Don't fragment
        MF:                              0 = Last fragment
    Fragment Offset:                     0
    Time to Live:                        63
    Protocol:                            TCP (6)
    Checksum:                            0x8958
    Source Address:                      192.168.129.42
    Destination Address:                 192.168.128.12
TCP Header
    Source Port:                         5001
    Destination Port:                    54123
    Sequence Number:                     1200057494
    Acknowledge Number:                  3740514055
    Data Offset:                         32
    Control Bits:                        0x10
        ACK                              Acknowledgment field significant
    Window Size:                         905
    Checksum:                            0x65b1
    Urgent Pointer:                      0
    Option Kind:                         No-operation (1)
    Option Kind:                         No-operation (1)
    Option Kind:                         Timestamp (8)
    Option Length:                       10
    Option Data:                         0x00
    Option Data:                         0x62
    Option Data:                         0x43
    Option Data:                         0x88
    Option Data:                         0x00
    Option Data:                         0x62
    Option Data:                         0x15
    Option Data:                         0xa1
PDU:
0000-0015: 45 00 00 34 2f e4 40 00 3f 06 89 58 c0 a8 81 2a     E..4/.@.?..X...*
0016-0031: c0 a8 80 0c 13 89 d3 6b 47 87 6c 96 de f3 b7 07     .......kG.l.....
0032-0047: 80 10 03 89 65 b1 00 00 01 01 08 0a 00 62 43 88     ....e........bC.
0048-0051: 00 62 15 a1                                         .b..
```

ymasmoudi committed:
[agw][liagentd] implement liagentd service

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

ymasmoudi commented:
@koolzz @ardzoht could you look at this please ? thx

ymasmoudi committed:
[liagent] rework state cleanup and fix agw setup

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

ymasmoudi committed:
[agw][liagent] rework unit test

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

ardzoht commented:
> @koolzz @ardzoht could you look at this please ? thx

Can you please mark the PR ready for review? This way it will add also the relevant code owners for each part of your PR for review. Thanks!

ardzoht commented:
> > @koolzz @ardzoht could you look at this please ? thx
> 
> Can you please mark the PR ready for review? This way it will add also the relevant code owners for each part of your PR for review. Thanks!

Also it looks there are some conflicts, please rebase and fix so you can update the PR for review.

koolzz commented on lte/gateway/c/li_agent/CMakeLists.txt:
```diff
@@ -15,7 +15,7 @@ PROJECT(LIAgent)
 include($ENV{MAGMA_ROOT}/orc8r/gateway/c/common/CMakeProtoMacros.txt)
 
-if (NOT BUILD_TESTS)
+if (BUILD_TESTS)
```
pretty sure its (as I've shamelessly copied sessiond makefile)
if (NOT BUILD_TESTS)

themarwhal replied:
I think we can get rid of this condition. I had to put this in on SessionD because one of the unit tests crashes with asan enabled. If these tests have an issue with asan, we can file an issue to fix it later though. If we need to keep the condition, it should be `NOT BUILD_TESTS`

koolzz commented on lte/gateway/c/li_agent/MobilitydClient.h:
```diff
@@ -0,0 +1,62 @@
+/*
+ * gRPC client for MobilityService
+ */
+class MobilitydClient : public GRPCReceiver {
```
Should we use the async model as in the removed DirectorydClient?

ymasmoudi replied:
While in sync mode, only the first packet will trigger a request to mobilityd. Moving to async calls will introduce more issues,
- while waiting for mobilityd to respond for the first packet, other packets from the same session will be processed triggering more grpc requests.
- if mobilityd returns a subscriber ID for both src and dst IP. we will have a race condition to access the copied packet.
- If mobilityd does not find both src and dst IP, we will have a memory leak unless we introduce a global variable to sync the cleanup.

Anyway, I changed mobilityd to use async model but kept the call in sync model until we discuss this further.

koolzz commented on lte/gateway/c/li_agent/InterfaceMonitor.cpp:
```diff
@@ -11,69 +11,64 @@
+int InterfaceMonitor::start_capture() {
+  int ret;
+  while (true) {
+    ret = pcap_dispatch(
```
nice

koolzz reviewed (commented):
Overall pretty good looking, can you expand the unit test coverage I think that would help understand whats going on.
Additionally if you could expand the PR description a bit with whats going on that would make it easier to review.

Also the nprobe patch is super cool

ymasmoudi commented:
> > @koolzz @ardzoht could you look at this please ? thx
> 
> Can you please mark the PR ready for review? This way it will add also the relevant code owners for each part of your PR for review. Thanks!

there is a dependancy on PR7000 (pending merge). will ping you shortly after.

ymasmoudi committed:
change mobilityd to async implementation + fix headers and rebase

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

themarwhal commented on lte/gateway/c/li_agent/CMakeLists.txt:
```diff
@@ -93,12 +93,14 @@ link_directories(
   ${MAGMA_LIB_DIR}/service_registry)
 
 add_library(LI_AGENT
+    Utilities.h
```
Unrelated to this PR but we should file an issue to sanitize all the proto generation.
```
set(SMGR_LTE_CPP_PROTOS mconfig/mconfigs)generate_cpp_protos("${SMGR_LTE_CPP_PROTOS}" "${PROTO_SRCS}"  "${PROTO_HDRS}" ${LTE_PROTO_DIR} ${LTE_CPP_OUT_DIR})set(SMGR_CWF_CPP_PROTOS accounting context)generate_cwf_cpp_protos("${SMGR_CWF_CPP_PROTOS}" "${PROTO_SRCS}"  "${PROTO_HDRS}" ${CWF_PROTO_DIR} ${CWF_CPP_OUT_DIR})set(SMGR_ORC8R_GRPC_PROTOS directoryd)generate_grpc_protos("${SMGR_ORC8R_GRPC_PROTOS}" "${PROTO_SRCS}"  "${PROTO_HDRS}" ${ORC8R_PROTO_DIR} ${ORC8R_CPP_OUT_DIR})message("Proto_srcs are ${PROTO_SRCS}")
```
we shouldn't have to generate any cwf / accounting protos

themarwhal replied:
https://github.com/magma/magma/pull/7359

ymasmoudi replied:
thanks @themarwhal

ymasmoudi committed:
[agw][liagent] store mconfig to avoid expensive calls + fix unit test

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

ymasmoudi commented:
> Overall pretty good looking, can you expand the unit test coverage I think that would help understand whats going on.
> Additionally if you could expand the PR description a bit with whats going on that would make it easier to review.
> 
> Also the nprobe patch is super cool

I will be updating the description and adding more test cases tomorrow.

themarwhal commented:
@ymasmoudi Do these unit tests pass with asan enabled? If it does, we should just enable it for sanity.

ymasmoudi committed:
[orc8r][nprobe] add unit tests

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

ymasmoudi commented:
> @ymasmoudi Do these unit tests pass with asan enabled? If it does, we should just enable it for sanity.

@themarwhal Yes, they do. I just added another two test cases and updated the description btw. (will enable it tomorrow).

ymasmoudi committed:
[orc8r][nprobe] fix lint issues and enable asan for test builds

Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>
